File: apps/docker-server/cmd/renewlet/online_icon_sources.go
package main

const (
	appStoreOnlineIconSource = "appStore"
	appStoreStorefrontUS     = "us"
	appStoreStorefrontCN     = "cn"
)

var (
	// 顺序和 shared APP_STORE_STOREFRONTS 保持一致；它同时影响请求顺序、缓存粒度和双区结果 tie-break。
	appStoreSupportedStorefronts = []string{appStoreStorefrontUS, appStoreStorefrontCN}
	appStoreDefaultStorefronts   = []string{appStoreStorefrontUS}
)

type onlineIconSourceSetting struct {
	Enabled     bool     `json:"enabled"`
	Storefronts []string `json:"storefronts"`
}

type onlineIconSourceSettings map[string]onlineIconSourceSetting

type onlineIconSourceSettingPatch struct {
	Enabled     *bool     `json:"enabled"`
	Storefronts *[]string `json:"storefronts"`
}

func defaultOnlineIconSourceSettings() onlineIconSourceSettings {
	return onlineIconSourceSettings{
		appStoreOnlineIconSource: {Enabled: true, Storefronts: cloneStringSlice(appStoreDefaultStorefronts)},
	}
}

// storefronts 是 Apple 请求放大开关；只允许固定小集合，空列表不能被当成“关闭来源”的隐式语义。
func normalizeAppStoreStorefronts(storefronts []string) ([]string, bool) {
	if len(storefronts) == 0 {
		return nil, false
	}
	selected := map[string]struct{}{}
	for _, storefront := range storefronts {
		if storefront != appStoreStorefrontUS && storefront != appStoreStorefrontCN {
			return nil, false
		}
		if _, ok := selected[storefront]; ok {
			return nil, false
		}
		selected[storefront] = struct{}{}
	}
	out := make([]string, 0, len(selected))
	for _, storefront := range appStoreSupportedStorefronts {
		if _, ok := selected[storefront]; ok {
			out = append(out, storefront)
		}
	}
	return out, true
}

func appStoreStorefrontsOrDefault(storefronts []string) []string {
	if normalized, ok := normalizeAppStoreStorefronts(storefronts); ok {
		return normalized
	}
	// 兜底只用于旧 settings 缺字段或测试空值；正式写入空数组会在 patch decoder 直接失败。
	return cloneStringSlice(appStoreDefaultStorefronts)
}

func cloneStringSlice(value []string) []string {
	return append([]string(nil), value...)
}

File: packages/shared/src/online-icon-sources.ts
export const ONLINE_ICON_PROVIDERS = ["appStore"] as const;
// 地区列表是跨 Docker/Worker/前端的请求放大契约；新增地区前必须同步限流、缓存 key 和 UI 文案。
export const APP_STORE_STOREFRONTS = ["us", "cn"] as const;

export type OnlineIconProvider = (typeof ONLINE_ICON_PROVIDERS)[number];
export type AppStoreStorefront = (typeof APP_STORE_STOREFRONTS)[number];
export const DEFAULT_APP_STORE_STOREFRONTS = ["us"] as const satisfies readonly AppStoreStorefront[];

/** 在线图标来源只控制手动搜索入口；它们没有内置 SVG 的变体和索引刷新语义。 */
export interface OnlineIconSourceSetting {
  enabled: boolean;
  storefronts: AppStoreStorefront[];
}

export type OnlineIconSourceSettings = Record<OnlineIconProvider, OnlineIconSourceSetting>;
export type OnlineIconSourceSettingsPatch = Partial<Record<OnlineIconProvider, Partial<OnlineIconSourceSetting>>>;
type LooseOnlineIconSourceSettingsPatch = Partial<Record<OnlineIconProvider, {
  enabled?: boolean | undefined;
  storefronts?: readonly AppStoreStorefront[] | undefined;
} | undefined>>;

export const DEFAULT_ONLINE_ICON_SOURCES: OnlineIconSourceSettings = {
  appStore: { enabled: true, storefronts: [...DEFAULT_APP_STORE_STOREFRONTS] },
};

/** App Store storefront 顺序是跨端排序契约；同时勾选时固定 US 优先，避免结果闪动。 */
export function normalizeAppStoreStorefronts(
  storefronts: readonly AppStoreStorefront[] | undefined,
): AppStoreStorefront[] {
  // 空值只服务历史缺字段读取兜底；写入边界必须由 Zod/Go strict decoder 拒绝空数组。
  const selected = new Set(storefronts && storefronts.length > 0 ? storefronts : DEFAULT_APP_STORE_STOREFRONTS);
  return APP_STORE_STOREFRONTS.filter((storefront) => selected.has(storefront));
}

/** 合并设置 patch 时按来源局部覆盖，避免保存 App Store 开关时影响后续在线来源。 */
export function mergeOnlineIconSourceSettings(
  base: OnlineIconSourceSettings = DEFAULT_ONLINE_ICON_SOURCES,
  patch?: OnlineIconSourceSettingsPatch,
): OnlineIconSourceSettings {
  return Object.fromEntries(ONLINE_ICON_PROVIDERS.map((provider) => [
    provider,
    {
      ...base[provider],
      ...patch?.[provider],
      storefronts: normalizeAppStoreStorefronts(patch?.[provider]?.storefronts ?? base[provider].storefronts),
    },
  ])) as OnlineIconSourceSettings;
}

/** 清理未知在线来源和 UI 临时字段，保持 settings JSON 只保存稳定产品契约。 */
export function cleanOnlineIconSourceSettingsPatch(
  patch?: LooseOnlineIconSourceSettingsPatch,
): OnlineIconSourceSettingsPatch | undefined {
  if (!patch) return undefined;
  // 这里用于读取历史设置和合并已校验 patch，不能替代写入 schema；非法空数组必须在 API 边界失败。
  const entries = ONLINE_ICON_PROVIDERS.flatMap((provider) => {
    const value = patch[provider];
    if (!value) return [];
    const cleanValue: Partial<OnlineIconSourceSetting> = {};
    if (value.enabled !== undefined) cleanValue.enabled = value.enabled;
    if (value.storefronts !== undefined) cleanValue.storefronts = normalizeAppStoreStorefronts(value.storefronts);
    return Object.keys(cleanValue).length > 0 ? [[provider, cleanValue] as const] : [];
  });
  return entries.length > 0 ? Object.fromEntries(entries) as OnlineIconSourceSettingsPatch : undefined;
}
